crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -5889,11 +5889,19 @@ impl<'db> TypeInferenceBuilder<'db> {
         self.infer_binary_expression_type(binary.into(), false, left_ty, right_ty, *op)
             .unwrap_or_else(|| {
                 if let Some(builder) = self.context.report_lint(&UNSUPPORTED_OPERATOR, binary) {
-                    builder.into_diagnostic(format_args!(
+                    let mut diag = builder.into_diagnostic(format_args!(
                         "Operator `{op}` is unsupported between objects of type `{}` and `{}`",
                         left_ty.display(self.db()),
                         right_ty.display(self.db())
                     ));
+
+                    if op == &ast::Operator::BitOr && left_ty.into_class_literal().is_some()
+                        && right_ty.into_class_literal().is_some() && Program::get(self.db()).python_version(self.db())
+                            < PythonVersion::PY310
```

Comment:
I think we could probably add the note whenever `left_ty` and `right_ty` are both subtypes of `KnownClass::Type.to_instance(db)`? `is_assignable_to` would not be appropriate here because e.g. `Any` is assignable to `type`, but `is_subtype_of` should be safe